### ISSUE TITLE
Allow resource limits to be excluded from kserve pods

### ISSUE THREAD

**Issue by AKnowlesAtDigit** [kserve/kserve#2671]
/kind feature

Looking through the source code, it seems that KServe has a defaulting webhook that sets resource limits on pods. Our tech stack does not use limits and instead relies on the requests field exclusively to ensure services have enough resources. We've temporarily solved this with a kubemod ModRule to remove the limits when the pods are created.

**Comment by rehevkor5** [comment]
Indeed. CPU Limit can result in undesirable performance characteristics, so we prefer to avoid it. KServe should not silently add hard-coded resource values that weren't asked for. The current behavior appears to be undocumented, too.

**Comment by yuzisun** [comment]
Documentation aside I do not necessarily agree on this, why would you want a pod to consume unlimited resources ? each pod should have a max limit for either memory or cpu, it is essential in a multi tenancy environment. Why setting limit would affect your performance if you are setting according to your need.

**Comment by rehevkor5** [comment]
> why would you want a pod to consume unlimited resources ? each pod should have a max limit for either memory or cpu, it is essential in a multi tenancy environment. Why setting limit would affect your performance if you are setting according to your need.

Pods are at minimum are given a percentage of the machine's CPU cycles that correspond to their CPU Request. If there are excess unused cycles, then it may also be allowed to consume some of those cycles too (sometimes referred to a "burst"). So, to the best of my knowledge it is possible to achieve fair CPU sharing without using limits/quotas.

> each pod should have a max limit for either memory or cpu, it is essential in a multi tenancy environment.

I have no issue with memory limit, my comment was about CPU limit.

Your assumption that everyone using KServe has multi-tenant clusters may not be true. Also, there are various situations one might describe as "multi-tenant" but each has different characteristics. For example, if you're a cloud service provider running a serverless-function platform, you might run many customers' workloads in shared clusters, and you probably tie your billing to the CPU Limit. In that case, CPU Limit might be critically important. In contrast, for example, if you run a cluster with only corporate-internal workloads owned by cooperative teams and a shared cost structure, there may be no need to cap each Pod's CPU use. So, it is not always essential.

> Why setting limit would affect your performance if you are setting according to your need.

I think the "if you are setting according to your need" statement makes an assumption that every MLE/DS/engineer has the knowledge, understanding, experience, and time to discover and apply the optimal K8s configuration. My experience has been that this is very often not the case.

You can find numerous examples of people's experience with the negative performance impact of CPU Limit if you search for words such as "k8s cpu limit throttling". Indeed, due to the occasionally surprising negative effects of CPU quota, some companies have disabled cgroup quotas entirely. In short, there is a time slice of a certain duration during which CPU usage is counted. If the Pod reaches its limit during that time slice, it is suspended for at least as long as it takes for the next time slice to begin. When this throttling occurs, it can have a very negative impact on your app's latency, for obvious reasons. Also, such throttling frequently occurs even when the Pod's CPU utilization otherwise appears to be well below its Limit!

In short, KServe should not make this decision globally on behalf of its users. In my opinion, KServe should avoid silently modifying properties of the users' k8s specs (because it can be confusing & surprising) unless there is a very specific reason why it's necessary. Or, it should at least grant users control over that behavior.

**Comment by rehevkor5** [comment]
@yuzisun does that explanation help? Mostly it boils down to this:

K8s' behavior is that if you don't set a CPU Limit, you don't get a CPU Limit. By silently adding a CPU Limit, KServe's behavior is surprising (relative to K8s) and takes control away from the user.

**Comment by yuzisun** [comment]
@rehevkor5 thanks for the detailed explanation! I do not think throttling can happen when the cpu utilization is well below the limit especially in kubernetes environment where the scheduler makes sure that you never oversubscribe the total available cpus in the cluster by setting the cpu limit. The throttling issue was caused by the Linux kernel bug which had been fixed already.

**Comment by yuzisun** [comment]
By the way kserve does not enforce setting the cpu limit and you can remove the limits from the serving runtimes if you really want to do that, but in general kserve wants to enforce best practices while still offering the flexibility.

https://github.com/kserve/kserve/blob/master/config/runtimes/kserve-sklearnserver.yaml#L27

**Comment by rehevkor5** [comment]
> By the way kserve does not enforce setting the cpu limit and you can remove the limits from the serving runtimes

It does not seem that way to me (also, we already removed the default CPU Limit from our `ClusterServingRuntime` instances), but will have to verify. In the code, I see https://github.com/kserve/kserve/blob/master/pkg/apis/serving/v1beta1/inference_service_defaults.go#L36 and https://github.com/kserve/kserve/blob/c4978eed4f78c1ea210b3c686a16fff8187c42d6/pkg/apis/serving/v1beta1/inference_service_defaults.go#L60 which looks suspicious to my uneducated Golang eye.
